### ISSUE TITLE
feat: Support RSC & App Layouts

### ISSUE THREAD

**Issue by KATT** [trpc/trpc#3297]
### Describe the feature you'd like to request

We should make official support for Next 13 app layouts & RSC.

### Describe the solution you'd like to see

Being able to transparently use tRPC identically in RSC-components and in `"use client"`-components. Unsure how they can be hydrated.

This is very open-ended and hard to answer without tonnes of exploration. We've done some in https://github.com/trpc/next-13

### Describe alternate solutions

n/a

### Additional information

https://github.com/trpc/next-13

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

**Comment by toddgeist** [comment]
tldr; Have `createTRPCNext` return a Provider like the `creatTRPCReact` version, and have it follow the `srr:false` config

We have settled on using the Next13 `app` directory but not using any new data-fetching stuff yet. We are still trying to use TRPC for client-side data fetching. This seems to work so far, although getting set up is a bit tricky.  

The main idea is to think of it as using plain node or express on the back end and a separate react app on the front. There are two configs for TRPC, one for the Next API handler and one for the client. This works, but since we are running next, those client components may run on the server on the first-page load, which breaks stuff. The workaround is to only do the TRPC queries inside dynamically imported components with `ssr:false`

Al this works, but it is a bit of a pain, and I feel some tweaks could make this setup easier. 

And most importantly, I think this is a worthwhile goal. The ultimate expression of what TRPC needs to be in the new RSC paradigm is an important question, but it will take some time to figure out. But in the meantime, if we could make this approach easier, it would give more people an opportunity to try out TRPC in next13 and provide feedback.

I don't feel I know the codebase well enough to tackle a PR on my own, but perhaps with some guidance I could get something together.

Thought?

**Comment by AlanMorel** [comment]
Hello, I was wondering if there's been any progress on this issue.

With Next.js working towards stabilizing the app directory, more people will be switching over and with that wishing to use tRPC with it.

Cheers!

**Comment by spencerbart** [comment]
Cheering on this development! I'm excited to see it soon.

**Comment by ansh** [comment]
I stopped using the `app` directory because of this issue

**Comment by akomm** [comment]
I don't say that everything is fine and perfect, just try to figure out what we have right now and what **probably** works:

**Server:** 
The API itself isn't a problem, remained  the same within `pages/api`. 

**Client:**
I think you can make it work without breaking RSC and turning everything into a client component. Don't follow the client part of [usage with Next.js](https://trpc.io/docs/nextjs), instead follow that of [usage with React](https://trpc.io/docs/react).

Create a TrpcProvider as a client-component:
```ts
"use client"

import {QueryClient, QueryClientProvider} from "@tanstack/react-query"
import {httpBatchLink} from "@trpc/client"
import {useState} from "react"
import {trpc} from "~/utils/trpc"

export const TrpcProvider: React.FC<{children: React.ReactNode}> = p => {
  const [queryClient] = useState(() => new QueryClient())
  const [trpcClient] = useState(() =>
    trpc.createClient({
      links: [
        httpBatchLink({
          url: "http://localhost:3000/api/trpc"
        })
      ]
    })
  )
  return (
    <trpc.Provider client={trpcClient} queryClient={queryClient}>
      <QueryClientProvider client={queryClient}>
        {p.children}
      </QueryClientProvider>
    </trpc.Provider>
  )
}
```
You don't have to, but you can even use it at the root level of your layout `app/layout.tsx`, the content within the provider still be rendered as RSC:
```ts
// app/layout.tsx
import {TrpcProvider} from "~/utils/trpc"

export default function Layout({children}: {children: React.ReactNode}) {
  return (
    <TrpcProvider>
      <html>
        <head>{/* snip */}</head>
        <body>{children}</body>
      </html>
    </TrpcProvider>
  )
}
```

I've tested it and the content is rendered on the server. Its supposed to work this way. Actually its one of core concepts of the new next 13 RSC implementation. 

Now this is for using trpc on the client. So if I get it correct, the whole missing point currently is to be able to run it on the RSC as well, right? Or are there some "details" that make the approach above inefficient or bad in some way?

**Comment by akomm** [comment]
I think the following needs to be considered - the doc page is already there, even though the link to it is disabled:
https://beta.nextjs.org/docs/api-reference/file-conventions/route
*(feature might not be ready yet, otherwise they would have linked it)*

This will eliminate the need to use `pages/api` eventually.

**Comment by michaelhays** [comment]
[Route handlers](https://twitter.com/nextjs/status/1626341664402051072?t=7G2oxuvoFmIUeB8t7pJX2w&s=19) just dropped!

@KATT Any word about progress on this? I assume there will be some work to do with this announcement but I'd love to hear from the team on where things are at :)

**Comment by Fredkiss3** [comment]
@michaelhays I migrated my trpc apis to route handlers, and it was simple, you just have to use the `fetch` adapter instead of nextjs api routes adapter :

1. I created a file under the path `app/api/trpc/[trpc]/route.ts`
2. I copied the adapter code like instructed in the docs : 

```ts
import { fetchRequestHandler } from '@trpc/server/adapters/fetch';
import { createContext } from './context.ts';
import { appRouter } from './router.ts';

const handler = (request: Request) => {
  return fetchRequestHandler({
    endpoint: '/api/trpc',
    req: request,
    router: appRouter,
    createContext,
  });
}

export const GET = handler;
export const POST = handler;
```

Keep in mind that you would have to change your createContext function signature also.

**Comment by Fredkiss3** [comment]
It doesn't, what I did is just move the tRPC api endpoint from next api routes to route handlers.

**Comment by revmischa** [comment]
How would one get access to the current next-auth session in `createContext`?

**Comment by mysterybear** [comment]
@Fredkiss3 how are you then calling/querying that from the front-end components and bringing types through?

my invocation of an exported `trpc` created with `createTRPCNext` is giving me this

![image](https://user-images.githubusercontent.com/50906986/224853481-2890cdbd-3ef5-4dba-bfff-6c291d64f8ba.png)

```
Unhandled Runtime Error
TypeError: Cannot destructure property 'abortOnUnmount' of 'useContext(...)' as it is null.

Source
Object.useQuery$1 [as useQuery]
node_modules/@trpc/react-query/dist/createHooksInternal-de11647d.mjs (294:16)
eval
node_modules/@trpc/react-query/dist/createHooksInternal-de11647d.mjs (57:0)
Object.apply
node_modules/@trpc/server/dist/index-972002da.mjs (18:0)
```

maybe you're using the "usage with React" rather than "usage with Next" guidance as @akomm suggested above? I'll try that now also

**Comment by StringKe** [comment]
@akomm Can you share the whole project structure? I've encountered some similar problems and I'd like to know how to deal with it.

**Comment by akomm** [comment]
As I've explained here, don't use the next HOC, use the react approach:
https://github.com/trpc/trpc/issues/3297#issuecomment-1423905894

Judging by @mysterybear's comment he used HOC. And if you have the same problem, it will be for the same reason. Unless something changed in some recent update.

**Comment by StringKe** [comment]
@akomm There is a `trpc.Provider` in your case and I'm not quite sure where he came from, I checked `@trpc/client` and `@trpc/next` and it doesn't seem to expose the provider.

**Comment by akomm** [comment]
It comes from the page I've linked:
https://trpc.io/docs/react#2-create-trpc-hooks
```ts
// utils/trpc.ts
import { createTRPCReact } from '@trpc/react-query';
import type { AppRouter } from '../path/to/router.ts';

export const trpc = createTRPCReact<AppRouter>();
``

**Comment by AlanMorel** [comment]
I realize this is potentially a lot to ask of someone, but if somebody who has it working already could create a tiny barebones repo, I think it would clear up a lot of back and forth here

**Comment by devietti** [comment]
@AlanMorel I have been trying to port an existing Next 13 app, so I ended up assembling [a little example repo](https://github.com/devietti/trpc-next13-app). Only the client makes RPC calls, though, no RPCs from RSCs.

**Comment by stychu** [comment]
Any more movement on this @KATT ?? Seems like `app` dir is becoming stable with next minor release so it should be marked as "production" ready I suppose? Would be cool to have tRPC available with it to <3

**Comment by KATT** [comment]
Planning to work on this thursday and friday.

It's hard to design an API when the full story isn't there yet:

- Still don't know how to hydrate `"use client"` components on first SSR w/o wrapping it in a RSC-component which to me feels completely redundant
- Mutation API still missing

**Comment by KATT** [comment]
We might release something prefixed as `unstable_` soon

**Comment by Thinkscape** [comment]
> We might release something prefixed as `unstable_` soon

Do you anticipate something wildly different to the approach in https://github.com/trpc/next-13 ?
NGL it was a bit confusing at first, but I was able to stabilise it in my app against trpc10 and latest next.

I share your concern about the userland API that is a bit unergonomic ATM (on top of the server/client boundary nesting doll pattern)

**Comment by matannahmani** [comment]
any updates regarding the unstable release?

**Comment by 413n** [comment]
In Next.js `13.4.0` they released [Server Actions](https://nextjs.org/blog/next-13-4#server-actions-alpha) in Alpha.
I hope that this unlocks the works for tRPC to fully support the App Directory.
Looking at what tRPC did for Next with `/pages`, when it will also support `/app` it will be HUUGE!

**Comment by matannahmani** [comment]
@413n I saw that @KATT already started experimenting with it, I will also give it a spin tmr

**Comment by lirbank** [comment]
In case this helps anyone, you can use the documented Next.js setup (for the pages router) as-is with an app router based app, by exporting the provider from a non RSC file:

```ts
// providers.ts
"use client";

import { api } from "@/lib/trpc";

export const TrpcProvider = api.withTRPC(
  (props: React.PropsWithChildren) => props.children
) as React.ComponentType<React.PropsWithChildren>;
```

```ts
// layout.ts
import { TrpcProvider } from "./providers";

export default function RootLayout(props: { children: React.ReactNode }) {
  return (
    <html lang="en">
      <body>
        <TrpcProvider>{props.children}</TrpcProvider>
      </body>
    </html>
  );
}
```

Will not work with RSC etc, but if you just want to get tRPC going with the new app dir.

**Comment by c-ehrlich** [comment]
> In case this helps anyone, you can use the documented Next.js setup (for the pages router) as-is with an app router based app, by exporting the provider from a non RSC file:
> 
> Will not work with RSC etc, but if you just want to get tRPC going with the new app dir.

You can also (preferably) extract queries into functions, or use the caller to get data serverside, either to render in an RSC or to pass to a client component and stick into initialData for a ssr/hydration-like pattern. But better patterns are hopefully coming in the not too distant future.

**Comment by AWash227** [comment]
Echoing what @c-ehrlich said, I've successfully managed to *pre-fetch* data in my RSC and then put that data into a client component using `createCaller` on the main appRouter, and `initialData` within each client query call. Although I'm not sure this is best-practice for now, it seems to work well for my needs. Let me know if this works for you, and I'm happy to move this comment somewhere else if needed.

Here's how it works, we define our backend logic within a router (I've copied over a router that works from my application). 
```ts
export const myRouter = router({
  findMany: procedure.query(/*[...]*/),
  findFirst: procedure.query(/*[...]*/),
  create: procedure.mutation(/*[...]*/),
  update: procedure.mutation(/*[...]*/),
  delete: procedure.mutation(/*[...]*/)
});

```
Mounted the main appRouter in the app directory at `app/trpc/[trpc]/route.ts` like so:
https://trpc.io/docs/server/adapters/nextjs#route-handlers


```ts import { appRouter } from "@/lib/rpc/router";
import { fetchRequestHandler } from "@trpc/server/adapters/fetch";

const handler = (req: Request) =>
  fetchRequestHandler({
    endpoint: "/api/trpc",
    req,
    router: appRouter,
    createContext: () => ({}),
  });

export { handler as GET, handler as POST };
```

Then created client hooks and a server-side api that we can use throughout the app:
```ts
export const t = createTRPCReact<AppRouter>({/*[...]*/})
```
and for server-side
```ts
export const s = appRouter.createCaller({/*[...]*/});
```
Setup the provider for client side queries:
```ts
export const Provider = ({ children }: {children: React.ReactNode}) => {
  return (
    <t.Provider client={trpcClient} queryClient={queryClient}>
      <QueryClientProvider client={queryClient}>
          {children}
          <ReactQueryDevtools />
      </QueryClientProvider>
    </t.Provider>
  );
};

```

Finally, within an RSC we can prefetch the data like so (ignore all of the implementation details just look at the line containing `initialData`:
```ts
export default createPage(
  z.object({ params: Params.findMany, searchParams: SearchParams.findMany }),
  async ({ params, searchParams }) => {
    const { server } = getResourceFromUri(params.uri);
    const initialData = await server.api.findMany({});
                                  // ^ this is our createCaller() instance 

    return (
      <ResourceTable initialData={initialData} searchParams={searchParams} />
    );
  },
);
```

Then in the component be sure to pass it as initial data:

```ts
const { data, fetchNextPage, isFetching } = hooks.findMany.useInfiniteQuery(
                                         // ^ this is our createTRPCReact() instance
    searchParams,
    { initialData: { pageParams: [], pages: [initialData] } },
    /* ^ we are passing our initialData to a useInfiniteQuery() hook, 
       so we need to pass it in as if it was our first fetched page*/
  );
```

**Comment by Frost7994** [comment]
Are we any closer to stable version release for the App Dir @KATT?

**Comment by kimdanielarthur-cowlabs** [comment]
For using trpc on the server side in nextjs 13 app router I used the createTRPCProxyClient approach. This in the case where the trpc endpoints are served from another backend.

trpcServer.ts
`import { createTRPCProxyClient, httpBatchLink } from "@trpc/client";
import type { AppRouter } from "../../trpc-server-in-monorepo/src/trpcRouters";
export const trpcProxyClient = createTRPCProxyClient<AppRouter>({
	links: [
		httpBatchLink({
			url: TRPC_SERVER_URL,
		}),
	],
});
`

And using it for example to generate page metadata:
page.tsx
`import { trpcProxyClient } from "../../utils/trpcServer";

type Params = {
	slug: string;
};

type Props = {
	params: Params;
	searchParams: { [key: string]: string | string[] | undefined };
};

export async function generateMetadata({ params, searchParams }: Props, parent: ResolvingMetadata): Promise<Metadata | undefined> {
	return trpcProxyClient.endpoint.query({slug:params.slug}).then((response)=>{ return {title: response.title}})
}
`

This seems to work well, however is this a bad approach with drawbacks?

**Comment by bibblebabl** [comment]
> DARN, this thing hasn't been released to work with, and I've been stuck on it for so long that I even ended up reading API codes, LOL. I hope support for the app router is released as soon as possible; it would be a fantastic!

AFAIK there is a way to use it with App Router already. 

I don't know why TRPC team hasn't updated the doc with these examples, but it's actually working solutions you can adapt. 

[https://youtu.be/qCLV0Iaq9zU?si=-Cf4ZmpsgSDUzeJ7](https://youtu.be/qCLV0Iaq9zU?si=-Cf4ZmpsgSDUzeJ7)


[https://youtu.be/hp-z01yAaRU?si=qJehUg5cGeD3ECIL](https://youtu.be/hp-z01yAaRU?si=qJehUg5cGeD3ECIL)

[https://youtu.be/iHJuJdw1jEM?si=swarcGwsBpy2vtG_](https://youtu.be/iHJuJdw1jEM?si=swarcGwsBpy2vtG_)

**Comment by me-imfhd** [comment]
yeah i know this, but it uses createTRPCReact and not createTRPCNext, also it uses adatapter as fetch instead of next,
so next is basically limited to pages router due to createTRPCNext, it does not has that feature to be used in app router the way it is used in pages router, in the first first video i guess the guy mentions this but not sure

**Comment by Azgaar** [comment]
Yes, tried it, it doesn't work fine. The session is getting lost

**Comment by giomogna** [comment]
There is a PR open on the T3 Stack using this approach, seems to be working flawlessly

https://github.com/t3-oss/create-t3-app/pull/1567

**Comment by ajdiyassin** [comment]
Any progress on this, now that server actions are stable? [nextjs.org/blog/next-14](https://nextjs.org/blog/next-14)

**Comment by natedunn** [comment]
@ajdiyassin Depends what you are asking for specifically. I am using tRPC with RSC's and App dir, but if you are asking specifically about out-of-the-box server actions, I'd keep an eye on #4567 reopening once they make progress.

**Comment by ian-os** [comment]
@ajdiyassin We're using the ts-oss repo recommendation from @giomogna and it's working well. I'd suggest starting there.

**Comment by shammalie** [comment]
T3 app router seems to work well, however, i'm struggling with getting subscription websocket server to work because of createTRPCContext uses next headers. This causes header invariant warnings because of passing the context and router to applyWSSHandler.

**Comment by thacoon** [comment]
@shammalie agree, [t3 experimental app router](https://create.t3.gg/) works fine, but I also can't get subscriptions running. If anyone has an example using the t3-stack as a base and could share it, that would be fantastic :)

**Comment by ItzDerock** [comment]
@thacoon, this might not be exactly what you're looking for, but for websockets, I ended up using a custom router on top of a t3 app scaffolded using the experimental app router feature. It's not the most ideal solution (I'm not sure if you can deploy this to vercel), but it works for my use case and depending on yours, it might work too.

<details>
<summary>my server entrypoint</summary>

```js
const app = next({
  dev: env.NODE_ENV !== "production",
  hostname: env.HOSTNAME,
  port: env.PORT,
  customServer: true,
});

await app.prepare();

const getHandler = app.getRequestHandler();
const upgradeHandler = app.getUpgradeHandler();

const server = createServer((req, res) => {
  // routes starting with /api/trpc are handled by trpc
  if (req.url?.startsWith("/api/trpc")) {
    const path = new URL(
      req.url.startsWith("/") ? `http://127.0.0.1${req.url}` : req.url,
    ).pathname.replace("/api/trpc/", "");

    return void nodeHTTPRequestHandler({
      path,
      req,
      res,
      router: appRouter,
      createContext: ({ req }) => {
        return createTRPCContext({
          req,
          res,
        });
      },
    });
  }

  getHandler(req, res).catch((error) => {
    logger.error(error);
    res.statusCode = 500;
    res.end("Internal Server Error");
  });
});

// create the websocket server
const wss = new WebSocketServer({ noServer: true });
const trpcHandler = applyWSSHandler({
  wss,
  router: appRouter,
  createContext: ({ req }) => {
    return createTRPCContext({
      req,
    });
  },
});

process.on("SIGTERM", () => {
  logger.warn("SIGTERM received, shutting down...");
  trpcHandler.broadcastReconnectNotification();
  server.close(() => {
    process.exit(0);
  });
});

// handle the upgrade
server.on("upgrade", (req, socket, head) => {
  // send trpc requests to the trpc server
  if (req.url?.startsWith("/api/trpc")) {
    wss.handleUpgrade(req, socket, head, (ws) => {
      wss.emit("connection", ws, req);
    });
  } else {
    void upgradeHandler(req, socket, head);
  }
});

server.listen(env.PORT, env.HOSTNAME, () => { ... });
```
</details>

**Comment by thacoon** [comment]
@ItzDerock thanks for your quick reply.

I also just made it worked (for development) based on [https://github.com/trpc/examples-next-prisma-websockets-starter](https://github.com/trpc/examples-next-prisma-websockets-starter), this example is not using the app router but it helped me solving it.

<details><summary>Details</summary>
<p>
Project is based on the t3-stack using app router.

My _src/server/wsServer.ts_ looks like this now (copied from the example above):
```ts
import {
  applyWSSHandler,
  type CreateWSSContextFnOptions,
} from '@trpc/server/adapters/ws';
import { WebSocketServer } from 'ws';
import { appRouter } from '~/server/api/root';
import { type CreateNextContextOptions } from '@trpc/server/adapters/next';

const wss = new WebSocketServer({
  port: 3001,
});

export const createContext = async (
  opts: CreateNextContextOptions | CreateWSSContextFnOptions
) => {
  return {};
};

const handler = applyWSSHandler({
  wss,
  router: appRouter,
  createContext,
});

wss.on('connection', (ws) => {
  console.log(`➕➕ Connection (${wss.clients.size})`);
  ws.once('close', () => {
    console.log(`➖➖ Connection (${wss.clients.size})`);
  });
});
console.log('✅ WebSocket Server listening on ws://localhost:3001');

process.on('SIGTERM', () => {
  console.log('SIGTERM');
  handler.broadcastReconnectNotification();
  wss.close();
});
```

The types for _createContext_ is wrong right know and I need to fix it but it works, after some hours of trial and error.

And my `src/server/trpc/react.tsx`:
```ts
'use client';

import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import {
  createWSClient,
  loggerLink,
  unstable_httpBatchStreamLink,
  wsLink,
} from '@trpc/client';
import { createTRPCReact } from '@trpc/react-query';
import { useState } from 'react';

import { type AppRouter } from '~/server/api/root';
import { getUrl, transformer } from './shared';

export const api = createTRPCReact<AppRouter>();

export function TRPCReactProvider(props: {
  children: React.ReactNode;
  cookies: string;
}) {
  const [queryClient] = useState(() => new QueryClient());

  const [trpcClient] = useState(() =>
    api.createClient({
      transformer,
      links: [
        loggerLink({
          enabled: (op) =>
            process.env.NODE_ENV === 'development' ||
            (op.direction === 'down' && op.result instanceof Error),
        }),
        wsLink({
          client: createWSClient({
            url: `ws://localhost:3001`,
          }),
        }),
        unstable_httpBatchStreamLink({
          url: getUrl(),
          headers() {
            return {
              cookie: props.cookies,
              'x-trpc-source': 'react',
            };
          },
        }),
      ],
    })
  );

  return (
    <QueryClientProvider client={queryClient}>
      <api.Provider client={trpcClient} queryClient={queryClient}>
        {props.children}
      </api.Provider>
    </QueryClientProvider>
  );
}
```

And similar to the example I have this in my *package.json*:
```json
{
  ...
  "scripts": {
    ...
    "dev:wss": "dotenv tsx watch src/server/wsServer.ts --tsconfig tsconfig.json",
    "dev:next": "next dev",
    "dev": "run-p dev:*",
    ...
  },
 ...
}
```
</p>
</details>

**Comment by shammalie** [comment]
@ItzDerock I tried your example below but couldn't get it to work, could you have a look?
https://github.com/shammalie/playground-t3-app/blob/main/src/server/prodServer.ts

**Comment by thacoon** [comment]
<details><summary>Off topic</summary>
<p>

> @thacoon I've tried to reuse the context in T3 app router, but i get errors with next-auth, since the app router context collects the session via getServerAuthSession(), this proceeds with the following: ProviderX is not a function, ProviderX being Discord in this example, or any provider. I'm realativly new to typescript so it could be a config issue, but i beleive the function error comes from the module augmentation because the compiler can't find next-auth module?
> 
> The error above comes from running the WSServer.
> 
> Here's the repo i use to test weird and wacky things with T3 stack. https://github.com/shammalie/playground-t3-app

In your local .env I think you need to assign DISCORD_CLIENT_ID, DISCORD_CLIENT_SECRET some value (at least just something), I think this is the error you get if they are empty.

</p>
</details>

**Comment by xsjcTony** [comment]
Do we have any ETA for the official release (even unstable) for this please? @KATT

**Comment by KATT** [comment]
There is an experimental release. I will start adopting RSC myself in prod over the next few months and it'll get more love gradually throughout that - some other core members are actively looking at this too.

Please do not `@` OSS-maintainers to do things for you - you're free to sponsor the issue using the link above and/or try to contribute to the project and/or suggest good APIs how it'd look like. If you're not contributing financially nor with work, please be patient with us when we do mostly work on this mostly for free.

**Comment by michaelhays** [comment]
Nothing but love and respect for everything you do here Alex!

If anything, I do understand the desire to stay in the know for these things that impact the future of the library so much, just for an understanding of where they are in the pipeline. I've been trying to follow along by checking various branches and PRs here and in T3 every couple of weeks, but haven't been able to figure much out.

Even a short message from time-to-time such as "This isn't stable enough in Next.js, we're going to wait at least several months before releasing something" or even "We're working on other things for now" is always much appreciated :)

**Comment by rwieruch** [comment]
@KATT where can we find the experimental release so that we can test it too? :) Thanks for all your hard work! 💯

**Comment by TzviPM** [comment]
I would be willing to build out a POC for this if it hasn't been done yet. My proposed approach would be to move the react query client from context into the trpc client or a wrapper. This would allow us to provide react query's prefetchQuery functionality inside of server components. We already seem to have this functionality, it's just accessed with useContext, which won't work in a server component. Thoughts?

**Comment by sneko** [comment]
@ItzDerock @thacoon you were mentioning solution to launch the websocket server, I'm able to. But did you succeed using a `trpc.myCustomName.useSubscription()` client into a `appDir` component? On my side it does nothing.

I'm curious if you could point to the repository working for you.

More details on my tests: https://github.com/trpc/trpc/discussions/5488

Thank you,

**Comment by prashantbhudwal** [comment]
For folks who want to use tRPC with app router and seek examples, this the best in-prod codebase out there - 
https://github.com/documenso/documenso.git

Till we get something stable, this a great stopgap.

**Comment by iantanwx** [comment]
> For folks who want to use tRPC with app router and seek examples, this the best in-prod codebase out there - https://github.com/documenso/documenso.git
> 
> Till we get something stable, this is a great stopgap.

Can confirm documenso solution is a solid one. I have used it to bootstrap a new SaaS, albeit with v11 beta.

**Comment by vertis** [comment]
As far as I can tell from the [github.com/documenso/documenso.git](https://github.com/documenso/documenso.git) repo this is just doing client side support. Doesn't (that I could see) have any RSC, SSR elements. 

FWIW It's pretty similar to the setup I'm currently using, which I did not get directly from that repo but is possibly inspired by. I can confirm that it does work but not for prefetching (or equiv)

**Comment by prashantbhudwal** [comment]
@vertis documenso uses data fetching with async functions in server components, and when they mutate the data using trpc, they just call `router.refresh()` after the mutation. That way, you get SSR and mutation capabilities from trpc. So.

In server components
- fetch data with async functions/fetch and pass it down to client components

In client components
- use this data directly (without adding it as initial values to query)
- call router.refresh() after mutation is done using tRPC/query

Again, everything is a stopgap.

**Comment by musjj** [comment]
There is a lot of info on this thread and I'm kinda lost. But just to get it straight, is the best approach of using trpc with the App Router is to use `createTRPCReact` instead of `createTRPCNext`?

This seems to be the approach used by the repo examples I've gathered:

- https://github.com/jherr/trpc-on-the-app-router/blob/main/src/app/_trpc/client.ts
- https://github.com/documenso/documenso/blob/2fa-input/packages/trpc/react/index.tsx

Am I getting this right?

**Comment by prashantbhudwal** [comment]
@musjj The repo mentioned there has been archived. Here is the latest one with the official trpc roadmap for rsc support.

https://github.com/trpc/examples-next-app-dir

**Comment by luixo** [comment]
The important caveat one may encounter while implementing tRPC with RSC & App Router with protected routes is that making api calls in SSR client components needs cookies and there's no simple way to get them (RSC gets `next/headers`, CSR pass them implicitly) but to pass them to the client component from RSC which present a vulnerability of XSS attack.

[Hacky guys](https://github.com/apollographql/apollo-client-nextjs/issues/85#issuecomment-1753442277) from Apollo found [a way to work around it](https://github.com/apollographql/apollo-client-nextjs/issues/85#issuecomment-1877345572): they encrypt them in RSC and decrypt in SSR (while still having access to server env variables) and not able to decrypt them in CSR.

This is the first place I looked up for this issue, so I decided to leave it here for the future investigators :)

**Comment by musjj** [comment]
@luixo 

> making api calls in SSR client components needs cookies and there's no simple way to get them

Apparently it works if you prefetch it: 
https://github.com/t3-oss/create-t3-app/issues/1765
https://github.com/t3-oss/create-t3-app/pull/1932

**Comment by twillhorn** [comment]
This mostly works:
https://trpc.io/docs/client/react/server-components

But it's fragile:
- If you prefetch, and use useSuspenseQuery, things work alright. useSuspenseQuery will suspend until the prefetched data is available, and from what I can tell hydration is working (no query happening on the client)
- But if you fail to prefetch, during server rendering useSuspenseQuery will trigger a "client" fetch, and it will be missing auth headers

I see it is possible to override headers() in the client creation, but something feels wrong about server to server fetch like this (I guess https://github.com/vercel/next.js/discussions/60640 is inspired by this problem).

What I'd like to do for the time being is just disable queries during SSR (if they weren't prefetched), and leave the job to the client. So I'm looking for an option on createClient to do this.

Is is possible? Is there a hook somewhere to tell the trpc client "never start a query on the server"?

**Comment by juliusmarminge** [comment]
> So I'm looking for an option on createClient to do this.

`useQuery` doesn't fetch on server, `useSuspenseQuery` does. I think this is something you want to customize on the RQ layer
